### PR TITLE
Feature/chat 81 api doc finalize

### PR DESCRIPTION
**Pull request by MapleThanh**

### CODE REVIEW

File: documentation/chatmeleon-swagger.yaml [hunk 1]
```diff
@@ -248,14 +399,18 @@ components:
           type: string
           format: date-time
           description: The timestamp when the message was created.
+        seenIds:
+          type: array
+          items:
+            type: string
+            pattern: "^[0-9a-f]{24}$"
```
Comment (WataNekko) [inline documentation/chatmeleon-swagger.yaml:406]: What is this field? Maybe add description?

File: documentation/chatmeleon-swagger.yaml [hunk 2]
```diff
@@ -228,7 +326,60 @@ components:
...
+          type: array
+          items:
+            type: string
+            format: uuid
+            description: List of message IDs the user has seen
 
+    Account:
+      type: object
+      properties:
+        id:
```
Comment (WataNekko) [inline documentation/chatmeleon-swagger.yaml:345]: Does the API client need to know about this schema? Do we use this in our API now or in the future?

**Review by WataNekko — CHANGES_REQUESTED** [review]

File: documentation/chatmeleon-swagger.yaml (hunk 2 above)
Comment (MapleThanh) [reply documentation/chatmeleon-swagger.yaml:345]: Yea we dont need the Account schema, we can remove it

File: documentation/chatmeleon-swagger.yaml (hunk 1 above)
Comment (MapleThanh) [reply documentation/chatmeleon-swagger.yaml:406]: Okay, ill add one

File: documentation/chatmeleon-swagger.yaml [hunk 3]
```diff
@@ -181,13 +196,96 @@ paths:
...
+        - bearerAuth: []
+      responses:
+        "200":
+          description: A message record.
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: "#/components/schemas/Message"
```
[the remark this answers is not in the document]
Comment (MapleThanh) [reply documentation/chatmeleon-swagger.yaml:247]: yea its an error, it should only be one Message object

File: documentation/chatmeleon-swagger.yaml [hunk 4]
```diff
@@ -51,43 +51,66 @@ paths:
...
               schema:
-                $ref: "#/components/schemas/Conversation"
+                type: object
+                properties:
+                  conversationId:
+                    type: string
+                    pattern: "^[0-9a-f]{24}$"
+                  messagesId:
+                    type: string
+                    pattern: "^[0-9a-f]{24}$"
```
[the remark this answers is not in the document]
Comment (MapleThanh) [reply documentation/chatmeleon-swagger.yaml:92]: yes

File: documentation/chatmeleon-swagger.yaml [hunk 5]
```diff
@@ -181,13 +196,96 @@ paths:
...
+                body:
+                  type: string
+                image:
+                  type: string
+                conversationId:
+                  type: string
+                  pattern: "^[0-9a-f]{24}$"
+  /messages/{messagesId}:
+    get:
+      description: Retrieve a message infomation by its ID.
```
[the remark this answers is not in the document]
Comment (MapleThanh) [reply documentation/chatmeleon-swagger.yaml:225]: So for example, when we create a new non-group conversation with a new message, that post request returns the new conversation ID and the new message ID but not the entire schema for both of them (already discussed), so when the user has those two IDs, they would then need to fetch the convo based on the convo ID and the message based on the messs ID. We wouldn't need to do this if we were to return all data of the new convo and message objects.

File: documentation/chatmeleon-swagger.yaml [hunk 6]
```diff
@@ -181,13 +196,96 @@ paths:
...
+                type: array
+                items:
+                  $ref: "#/components/schemas/Message"
+        "401":
+          description: Unauthorized - JWT token is missing or invalid.
+        "403":
+          description: Unauthorized - User is unauthorized.
+  /user/messages:
+    get:
+      description: Retrieve all messages of the current user  with the support of cursor-based pagination. The page will be an array containing the newest messages to the oldest messgaes from the cursor. For example [mess1, mess2, mess3, mess4, mess5] with mess1 being the most recently sent message and mess5 being the oldest message.
```
[the remark this answers is not in the document]
Comment (MapleThanh) [reply documentation/chatmeleon-swagger.yaml:254]: Not for now we dont need it, so we can remove it

File: documentation/chatmeleon-swagger.yaml [hunk 7]
```diff
@@ -25,20 +25,20 @@ tags:
...
+      description: Retrieve a page of conversations associated with the current logged in user, with the support of cursor-based pagination. The page will be an array containing the newest conversation to the oldest conversation from the cursor. For example [convo1, convo2, convo3, convo4, convo5] with convo1 being the most recently active conversation and convo5 being the oldest active conversation.
       tags:
         - conversations
       summary: Get current user's conversations
       parameters:
         - name: pageSize
           in: query
-          description: The number of conversations in one page.
+          description: The number of conversation in one page. When this query is empty, a default value of 10 will be used. A positive pageSize will return pageSize number of conversation that is more recent than the cursor, for example [cursor = convo4.id; pageSize = 2] will return [convo2, convo3] not [convo3, convo2] because of the order mentioned above. A negative pageSize will do the opposite, [cursor = convo2.id; pageSize = -2] will return [convo3, convo4].
```
[the remark this answers is not in the document]
Comment (MapleThanh) [reply documentation/chatmeleon-swagger.yaml:35]: This is just how mongoDB works. But yea it is kinda counter-intuitive, so maybe later on we can like negate the pageSize value in the implementation controller?

[the remark this answers is not in the document]
Comment (MapleThanh) [reply documentation/chatmeleon-swagger.yaml:35]: Oh i see why now, just stupid mongoDB hehe. Let me fix it

[the remark this answers is not in the document]
Comment (WataNekko) [reply documentation/chatmeleon-swagger.yaml]: ye hehe

File: documentation/chatmeleon-swagger.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (WataNekko) [reply documentation/chatmeleon-swagger.yaml:227]: I think you're talking about a different route. I'm talking about this GET /messages/{messagesId}

[the remark this answers is not in the document]
Comment (MapleThanh) [reply documentation/chatmeleon-swagger.yaml:227]: Yea i am talking about `GET /messages/{messagesId}` cause we need to retrieve the message data and information by its ID when creating a new message everytimes we text (`POST /messages` only returns a message ID). An alternative would be re-fetching the whole page of messages of the conversation in `/conversations/{conversationId}/messages`, but I don't think that's really efficient, fetching the whole message page in the convo for every singular new message seems exhausting.

File: documentation/chatmeleon-swagger.yaml [hunk 8]
```diff
@@ -267,26 +347,78 @@ components:
...
-            pattern: "^[0-9a-f]{24}$"
-          description: An array of MongoDB _ids of users associated with this conversation. These user IDs can be used to reference the User model.
+            format: uuid
+            description: List of user IDs belong to the conversation
+        messageIds:
+          type: array
+          items:
+            type: string
+            format: uuid
+            description: List of message IDs belong to the conversation
```
Comment (WataNekko) [inline documentation/chatmeleon-swagger.yaml:384]: Remove this

Comment (WataNekko) [inline documentation/chatmeleon-swagger.yaml:384]: Remove this

**Review by WataNekko — CHANGES_REQUESTED** [review]

File: documentation/chatmeleon-swagger.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (WataNekko) [reply documentation/chatmeleon-swagger.yaml:227]: gotcha

File: documentation/chatmeleon-swagger.yaml [hunk 9]
```diff
@@ -181,13 +198,31 @@ paths:
           content:
             application/json:
               schema:
-                $ref: "#/components/schemas/Message"
+                type: object
+                properties:
+                  messageId:
+                    type: string
+                    pattern: "^[0-9a-f]{24}$"
```
Comment (WataNekko) [inline documentation/chatmeleon-swagger.yaml:205]: Add createdAt field to this response, to substitute the individual GET message route earlier

File: documentation/chatmeleon-swagger.yaml [hunk 10]
```diff
@@ -151,22 +175,15 @@ paths:
...
         "401":
           description: Unauthorized - JWT token is missing or invalid.
         "403":
-          description: Forbidden - User not allowed to retrieve the messages of the conversation.
-        "404":
-          description: Not Found - Conversation not found.
+          description: Unauthorized - User is unauthorized.
   /messages:
     post:
       description: Create a new message.
```
Comment (WataNekko) [inline documentation/chatmeleon-swagger.yaml:189]: Change this route path to `POST /conversations/{conversationId}/messages` to be consistent with earlier GET route. And of course, remove the conversationId field in the request body down below.

File: documentation/chatmeleon-swagger.yaml [hunk 11]
```diff
@@ -181,13 +198,31 @@ paths:
...
-              $ref: "#/components/schemas/Message"
+              type: object
+              properties:
+                body:
+                  type: string
+                image:
+                  type: string
+                conversationId:
+                  type: string
+                  pattern: "^[0-9a-f]{24}$"
```
Comment (WataNekko) [inline documentation/chatmeleon-swagger.yaml:224]: Remove this from the requestBody if already added to the path as a param.

**Review by WataNekko — CHANGES_REQUESTED** [review]
Last thing I swear hehe

**Review by WataNekko — APPROVED** [review]
lgtm. amazing good job boss. luv u

File: documentation/chatmeleon-swagger.yaml [hunk 12]
```diff
@@ -267,26 +320,77 @@ components:
...
+          format: uuid
+          description: Unique identifier for each relationship
+        type:
+          type: string
+          enum:
+            - FRIEND
+            - BLOCK
+            - FOLLOWING
+            - PENDING_FRIEND_REQUEST
+            - SENDING_FRIEND_REQUEST
```
Comment (XkhangNguyen) [inline documentation/chatmeleon-swagger.yaml:377]: About following, is there any enum value for the followed person? I think there should be a value for it.

**Review by XkhangNguyen — CHANGES_REQUESTED** [review]

File: documentation/chatmeleon-swagger.yaml (hunk 12 above)
Comment (MapleThanh) [reply documentation/chatmeleon-swagger.yaml:377]: It will in the future but for the current state of our app, this is sufficient

**Review by XkhangNguyen — APPROVED** [review]

**Pull request merged**